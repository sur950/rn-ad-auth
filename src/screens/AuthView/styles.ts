import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  closeBtn: {
    height: 38,
    width: 38,
    borderRadius: 19,
    backgroundColor: '#f7f7f7',
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'flex-end',
    borderWidth: 0.2,
    borderColor: '#c9c9c9',
    marginRight: 20,
    marginBottom: 10,
  },
  closeImg: {
    height: 18,
    width: 18,
  },
});

export default styles;
