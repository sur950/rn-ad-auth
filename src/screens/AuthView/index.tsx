import React from 'react';
import { Image, SafeAreaView, TouchableOpacity } from 'react-native';
import { WebView } from 'react-native-webview';
import constants from '../../utilities/constants';
import {
  getLoginUrl,
  getLogoutUrl,
  getParamValueFromUrl,
} from '../../utilities/helpers';
import { getAuthTokens, getUserDetails } from '../../services/auth';
import type { AuthViewProps } from '../../types';
import styles from './styles';

const AuthView = (props: AuthViewProps) => {
  const { config, action, onSuccess, onClose, closeModal } = props;

  /**
   *
   * @returns Url to be served to webview
   */
  const getWebUrl = (): string => {
    switch (action) {
      case 'login':
        return getLoginUrl(config);
      case 'logout':
        return getLogoutUrl(config.tenant_id);
      default:
        return '';
    }
  };

  /**
   *
   * - Handles Login by looking at navigation state changes of webview.
   * @param state
   * @returns
   */
  const handleLogin = async (state: any) => {
    const { url } = state;
    const isMSUrl = url.includes('https://login.microsoftonline.com/');

    // Returning here  by certain conditions to validate correct callback url
    if (!url.includes('session_state') || isMSUrl || state.navigationType) {
      return;
    }

    closeModal && closeModal();

    const code = getParamValueFromUrl('code', url);
    const data = {
      code,
      tenant_id: config.tenant_id,
      redirect_uri: config.redirect_uri,
      client_id: config.client_id as string,
      ...constants.DEFAULT_GET_TOKEN_PARAMS,
    };

    const tokenData = await getAuthTokens(data);
    const userData = await getUserDetails(tokenData?.access_token);

    const responseData = {
      code: 200,
      message: 'User logged in successfully.',
      ...tokenData,
      ...userData,
    };

    onSuccess(responseData);
  };

  /**
   *
   * @param state - Navigation State of webview
   */
  const handleLogout = (state: any) => {
    const { url } = state;
    if (!state.navigationType && url.includes('logoutsession')) {
      const responseData = {
        code: 200,
        message: 'User logged out successfully.',
      };
      onSuccess(responseData);
    }
  };

  /**
   * Runs when Url from Webview isChanged.
   */
  const onStateChange = (state: Object) => {
    switch (action) {
      case 'login':
        handleLogin(state);
        break;
      case 'logout':
        handleLogout(state);
        break;
      default:
    }
  };

  return (
    <SafeAreaView style={styles.container}>
      <TouchableOpacity onPress={onClose} style={styles.closeBtn}>
        <Image
          style={styles.closeImg}
          resizeMode="contain"
          source={require('../../assets/images/close.png')}
        />
      </TouchableOpacity>
      <WebView
        style={styles.container}
        source={{ uri: getWebUrl() }}
        onNavigationStateChange={onStateChange}
      />
    </SafeAreaView>
  );
};

export default AuthView;
