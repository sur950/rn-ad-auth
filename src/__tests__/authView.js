// __tests__ for AuthView Component

import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import AuthView from '../screens/AuthView';

const authViewProps = {
  config: {
    tenant_id: 'xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx',
    client_id: 'xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx',
    scope: 'openid',
  },
  action: 'login',
  onSuccess: () => {},
  onClose: () => {},
  closeModal: () => {},
};

describe('Authview test', () => {
  it(`should match the AuthView snapshot`, () => {
    const tree = renderer
      .create(<AuthView {...authViewProps} closeModal={() => {}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
