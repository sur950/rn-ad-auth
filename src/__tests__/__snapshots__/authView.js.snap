// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authview test should match the AuthView snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#f8f7f8",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f8f8f8",
        "borderBottomColor": "#d8d8d8",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "height": 44,
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#007aff",
            "fontSize": 18,
            "fontWeight": "400",
          }
        }
      >
        Cancel
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginLeft": 10,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "default": "",
          }
        }
        style={
          Object {
            "height": 12,
            "width": 12,
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 15,
            "fontWeight": "bold",
            "marginLeft": 5,
          }
        }
      >
        https://login.microsoftonline.com
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <RNCWebView
      cacheEnabled={true}
      injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
      injectedJavaScriptForMainFrameOnly={true}
      javaScriptEnabled={true}
      messagingEnabled={true}
      onContentProcessDidTerminate={[Function]}
      onHttpError={[Function]}
      onLoadingError={[Function]}
      onLoadingFinish={[Function]}
      onLoadingProgress={[Function]}
      onLoadingStart={[Function]}
      onMessage={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      source={
        Object {
          "uri": "https://login.microsoftonline.com/xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx/oauth2/v2.0/authorize/?tenant_id=xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx&client_id=xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx&scope=openid&response_type=code&response_mode=query&state=12345&nonce=678910&redirect_uri=https%3A%2F%2Fazure-ad-login-3606b.web.app%2F",
        }
      }
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#ffffff",
          },
          Object {
            "backgroundColor": "#f8f7f8",
            "flex": 1,
          },
        ]
      }
      textInteractionEnabled={true}
      useSharedProcessPool={true}
    />
  </View>
</RCTSafeAreaView>
`;
